plugins/ccgx/fu-ccgx-hpi.c

```diff
@@ -0,0 +1,1633 @@
+ *
+ * Returns: %TRUE for success
+*/
+static gboolean
+fu_ccgx_hpi_configure (FuDevice *device,
+		       CyHPIHandle *handle,
+		       guint8 *device_mode,
+		       GError **error)
+{
+	CyI2CConfig i2c_config;
+	guint8 mode;
+
+	g_return_val_if_fail (handle != NULL, FALSE);
+	g_return_val_if_fail (device_mode != NULL, FALSE);
+
+	handle->hpi_addr_byte = 1;
+	handle->num_of_ports = 1;
+
+	if (!fu_ccgx_i2c_get_config(device, &handle->i2c_handle, &i2c_config, error)) {
+		g_prefix_error (error, "fu_ccgx_hpi_configure  get config error: ");
+		return FALSE;
+	}
+
+	i2c_config.frequency = FU_CCGX_I2C_FREQ;
+	i2c_config.is_master = TRUE;
+
+	if (!fu_ccgx_i2c_set_config(device,&handle->i2c_handle,&i2c_config,error)) {
+		g_prefix_error (error, "fu_ccgx_hpi_configure -set config error: ");
+		return FALSE;
+	}
+
+	if (!fu_ccgx_hpi_get_device_mode(device,handle,&mode,error)) {
+		g_prefix_error (error, "fu_ccgx_hpi_configure -get device mode: ");
+		return FALSE;
+	}
+
+	*device_mode = mode;
+
+	if (mode & 0x80) {
+		handle->hpi_addr_byte = 2;
+	} else {
+		handle->hpi_addr_byte = 1;
+	}
+
+	if ((mode >> 2) & 0x03) {
+		handle->num_of_ports = 2;
+	} else {
+		handle->num_of_ports = 1;
+	}
+
+	handle->fw_mode = (FWMode)(mode & 0x03);
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_cmd_setup:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @dm_device Device Type for Device Manager
+ * @usb_inf_num USB interface number for I2C
+ * @slave_address Slave Address for I2C
+ * @error: a #GError or %NULL
+ *
+ * Setup PD I2C device using HPI interface
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_setup (FuDevice *device,
+		       CyHPIHandle *hpi_handle,
+		       DMDevice dm_device,
+		       guint16 usb_inf_num,
+		       guint8 slave_address,
+		       GError **error)
+{
+	CyI2CDeviceHandle *i2c_handle = &hpi_handle->i2c_handle;
+	guint8 device_mode = 0;
+	guint32	hpi_event = 0;
+	g_autoptr(GError) error_local = NULL;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+
+	i2c_handle->inf_num = usb_inf_num;
+	i2c_handle->slave_address = slave_address;
+
+	if (dm_device == DM_DEVICE_PD_I2C) {
+		i2c_handle->ep.bulk_out = PD_I2C_USB_EP_BULK_OUT;
+		i2c_handle->ep.bulk_in = PD_I2C_USB_EP_BULK_IN;
+		i2c_handle->ep.intr_in = PD_I2C_USB_EP_INTR_IN;
+	} else if(dm_device == DM_DEVICE_PD_I2CM) {
+		i2c_handle->ep.bulk_out = PD_I2CM_USB_EP_BULK_OUT;
+		i2c_handle->ep.bulk_in = PD_I2CM_USB_EP_BULK_IN;
+		i2c_handle->ep.intr_in = PD_I2CM_USB_EP_INTR_IN;
+	} else {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "i2c setup not supported device");
+		return FALSE;
+	}
+
+	if  (!fu_ccgx_hpi_configure (device, hpi_handle, &device_mode, error)) {
+		g_prefix_error (error, "i2c_setup error: ");
+		return FALSE;
+	}
+
+	if (fu_ccgx_hpi_get_event (device, hpi_handle, HPI_REG_SECTION_DEV, &hpi_event,
+			   HPI_CMD_SETUP_EVENT_WAIT_TIME_MS, &error_local)) {
```

Comment:
If you're not actually using error_local for anything (like printing a message), you should just send NULL for error.

Reply:
OK. I will set NULL for error here.